File: GeneticProgramming/genetic_text_python/population.py
import dna
import random as rd

def my_map(val, min_val, max_val, min_obj, max_obj):
  rangea = max_val - min_val
  rangeobj = max_obj - min_obj
  rangeobj /= rangea

  return (min_obj + val * rangeobj)




class Population:

  def __init__(self, pop_size, target, mutation_rate):
    self.dna = []
    self.matin_pool = []
    self.mutation_rate = mutation_rate
    self.pop_size = pop_size
    self.target = target


    #create the population
    for i in range(pop_size):
      self.dna.append(dna.Dna(len(target), target))

  def show(self):
    print(self.pop_size)
    print(self.target)

    for n in range(10):
      self.dna[n].show()

  def calculate_fitness(self):
    for adn in self.dna:
      adn.calc_fitness()


  def gen_candidates(self):
    self.matin_pool = []

    best_obj = max(self.dna, key = lambda adn: adn.fitness)
    best_fitness = best_obj.fitness

    # for adn in self.dna:
    #   if adn.fitness  > best_fitness:
    #     best_fitness = adn.fitness


    # print(f"best_fitnes: {best_fitness}")
    print("best till now:")
    best_obj.show()

    #normalize the fitness
    #there are two options to normalize
    # 1. divide all finess by the max fitness
    # 2. map the fitness between 0 and 1, based in the max fitness

    #normalize the fitness
    for adn in self.dna:
      adn.fitness = my_map(adn.fitness, 0, best_fitness, 0, 1)

    for adn in self.dna:
      # prob = adn.fitness * 100
      # if adn.fitness == 0: continue
      prob = my_map(adn.fitness, 0, best_fitness, 0, 1) * 100
      for i in range(int(prob)):
        self.matin_pool.append(adn)



    print(f"len mpool: {len(self.matin_pool)}")


  def reproduction(self):
    #create the new generation from the mating pool
    for i in range(len(self.dna)):
      # parentA = self.pick_one()
      # parentB = self.pick_one()
      parentA = self.matin_pool[rd.randint(0, len(self.matin_pool) - 1)]
      parentB = self.matin_pool[rd.randint(0, len(self.matin_pool) - 1)]
      child = parentA.crossover(parentB)
      child.mutate(self.mutation_rate)
      self.dna[i] = child


  def evaluate(self):
    for adn in self.dna:
      curr_len = 0
      if "".join(adn.dna) == self.target:
        return True
          
    return False
      

  def pick_one(self):
    rand = rd.random()
    i = 0
    while rand>0:
      rand -= self.dna[i].fitness
      i += 1
    i-=1
    # self.dna[i].show()

File: GeneticProgramming/genetic_cars/population.js
class Population {
  constructor(pop_size, st_x, st_y) {
    this.pop_size = pop_size;
    this.population = [];
    this.st_x = st_x;
    this.st_y = st_y;
    this.init_pop();
    this.matin_pool = [];
  }

  selection() {
    this.matin_pool = [];
    let max_fit = 0;
    for (let i = 0; i < this.pop_size; i++) {
      this.population[i].calculate_fitness();
      if (this.population[i].fitness > max_fit) {
        max_fit = this.population[i].fitness;
      }
    }

    //create mating pool
    for (let i = 0; i < this.pop_size; i++) {
      // print(this.population[i].fitness);
      let n = this.population[i].fitness / max_fit;
      
      for(let k = 0; k < int(n * 100); k++){
        this.matin_pool.push(this.population[i]);
      }
    }
  }
  
  
  reproduction(){
    for (let i = 0; i < this.pop_size; i++) {
      let parentA = random(this.matin_pool).dna;
      let parentB = random(this.matin_pool).dna;
      
      let child = parentA.crossover(parentB);
      child.mutation();
      
      this.population[i] = new Car(this.st_x, this.st_y, 10, 20, child);
    }
  }
  

  init_pop() {
    for (let i = 0; i < this.pop_size; i++) {
      // let car = new Car(200, 10, 10, 20);
      let car = new Car(this.st_x, this.st_y, 10, 20);
      this.population.push(car);
    }
  }

  show() {
    for (let i = 0; i < this.pop_size; i++) {
      this.population[i].update();
      this.population[i].show();
    }

  }
}
